bench/workgen/workgen.cxx

```diff
@@ -248,6 +281,231 @@ WorkloadRunner::start_table_idle_cycle(WT_CONNECTION *conn)
+    }
+
+    return 0;
+}
+
+/*
+ * This function will drop "dynamic_drop_count" tables every "dynamic_drop_period" seconds. It will
+ * also monitor the database size and stop dropping tables once the size goes below
+ * "dynamic_drop_target". It will restart table drops if the database size crosses
+ * "dynamic_drop_trigger".
+ */
+int
+WorkloadRunner::start_dynamic_drop(WT_CONNECTION *conn)
+{
+    WT_SESSION *session;
+    if (conn->open_session(conn, nullptr, nullptr, &session) != 0) {
+        THROW("Error opening a session.");
+    }
+
+    WT_DECL_RET;
+    ContextInternal *icontext = _workload->_context->_internal;
+    std::vector<std::string> tables_delete; // Track tables that are to be deleted
+    uint32_t db_size = get_dir_size_mb(_wt_home);
+
+    // Initially we start dropping tables if the database size is more than the drop trigger.
+    bool dropping = db_size > _workload->options.dynamic_drop_trigger;
+
+    while (!stopping) {
+        /*
+         * If we have been dropping, continue until we reach the drop target size. If we have not
+         * been dropping tables, begin to do so if the database size crosses the drop trigger.
+         */
+        db_size = get_dir_size_mb(_wt_home);
+        if (dropping) {
+            dropping = db_size > _workload->options.dynamic_drop_target;
+            if (!dropping) {
+                VERBOSE(*_workload,
+                  "Stopped dropping new tables. db_size now "
+                    << db_size << " MB has reached the drop target of "
+                    << _workload->options.dynamic_drop_target << " MB.");
+            }
+        } else {
+            dropping = db_size > _workload->options.dynamic_drop_trigger;
+            if (dropping) {
+                VERBOSE(*_workload,
+                  "Started dropping new tables. db_size now "
+                    << db_size << " MB has reached the drop trigger of "
+                    << _workload->options.dynamic_drop_trigger << " MB.");
+            }
+        }
+        if (!dropping) {
+            sleep(_workload->options.dynamic_drop_period);
+            continue;
+        }
+
+        for (int drops = 0; drops < _workload->options.dynamic_drop_count; ++drops) {
+            std::vector<std::string> drop_uris;
+            // The dynamic table data structures are protected by a mutex.
+            {
+                const std::lock_guard<std::mutex> lock(*icontext->_dyn_mutex);
+
+                // Walk the list and select the first that is not already in the delete list
+                for (auto const &it : icontext->_dyn_tint) {
+                    if (icontext->_dyn_table_runtime[it.second]._pending_delete) {
+                        continue;
+                    } else {
+                        VERBOSE(*_workload, "Marking pending removal for: " << it.first);
+                        icontext->_dyn_table_runtime[it.second]._pending_delete = true;
+                        tables_delete.push_back(it.first);
```

Comment:
Can we assert the uri `it.first` is not in the `tables_delete` structure?

Reply:
I am double-minded on this one and decided not to. I had thought about this earlier, but did not want to walk the whole list to check if tables_delete already contains the entry. If you have a different suggestion than walking the list each time, I am happy to look at it. The assert is less required than before as ```icontext->_dyn_table_runtime[it.second]._pending_delete``` is the same source of information on deletes now.

Reply:
Instead of walking, you can use the `find` method. It's probably walking the data structure under the hood.

Reply:
Ya, find does the walk, so I want to avoid doing that every iteration of delete. But I see merit in asserting. I still have to refine this code a bit when I switch to shared_mutex, so I will explore adding (more) asserts using that ticket, WT-10309. While trying to switch to shared mutex I observed some bugs and asserts could help in there.